### PR TITLE
Update: Add ImmutableJS

### PR DESCRIPTION
**Pull request by leeallen337**
## Why?

Issue https://github.com/smashingboxes/web-boilerplate/issues/64

After talks with @keeganbrown and @dkniffin we've reached the conclusion that we should add `ImmutableJS` to the boilerplate. We always end up using it on all our projects and having it already incorporated will help speed up scaffolding a new project.

## What Changed?
- Added the following NPM packages 
  - `immutable`
    - We always end up adding this so it should be in the boilerplate
  - `redux-immutable`
    - This allows our top level state to be an `Immutable` object rather than a regular JavaScript object
    - See [here](https://redux.js.org/docs/recipes/UsingImmutableJS.html#make-your-entire-redux-state-tree-an-immutablejs-object)
  - `react-immutable-proptypes`
    - So we can have prop types with Immutable objects
  - `redux-persist-immutable`
    - This is a wrapper around `redux-persist` to support ImmutableJS
    - See [here](https://github.com/rt2zz/redux-persist-immutable)
  - `chai-immutable`
    - Allow chai assertions for Immutable Collections
    - See [here](https://github.com/astorije/chai-immutable)
- Removed the following NPM package
  - `redux-persist`
    - Since we're using `redux-persist-immutable` which is just a wrapper for `redux-persist` we don't need this installed directly
- Updated all reducers, specs, etc. to support ImmutableJS

## Once Merged
I will update the issue referenced in the Why section

## More updates to come. Take a look at the Issue referenced in Why

### CODE REVIEW

File: modules/authentication/Interceptors.js [hunk 1]
```diff
@@ -15,10 +15,10 @@ class Interceptors {
     return this.store
       .getHydratedState()
       .then((state) => {
-        const tokenInfoFields = Object.keys(state.authentication.tokenInfo);
+        const tokenInfoFields = state.getIn(['authentication', 'tokenInfo']).keySeq().toArray();
```
Comment (dkniffin) [inline modules/authentication/Interceptors.js]: Any reason to do `keySeq` instead of just `keys`?

Comment (leeallen337) [reply modules/authentication/Interceptors.js]: @dkniffin Good question. I believe when setting this up on ROAMD I had initially copied the use of `keySeq` from somewhere. However, I've dug a little further into this and I'll outline what I've come up with.

From what I've gathered, using `keys` returns an `Iterator` while `keySeq` returns an `Iterable` and both have different APIs. They attempted to match the `Iterator` protocals (which can be found [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols)) for `keys` which gives you things like using `next` but for Immutable doesn't give you the `Sequence` API.

See some more info in the links below:
* https://github.com/facebook/immutable-js/issues/529
* https://stackoverflow.com/questions/35623347/immutable-js-get-keys-from-map-hash

But basically if you use `keys` you can't then also do an `toArray` at the end.

state

File: modules/authentication/reducer.js [hunk 2]
```diff
@@ -1,89 +1,84 @@
...
-    };
+    return state.withMutations((map) => {
+      map.set('email', null);
+      map.set('error', null);
+      map.set('id', null);
+      map.set('isActive', false);
+      map.set('name', null);
+      map.set('uid', null);
+      map.set('tokenInfo', Immutable.fromJS({}));
+    });
```
Comment (OzoneTank) [inline modules/authentication/reducer.js]: it looks like you've added `error` and `isActive` I'm assuming they were just missing before.

File: modules/authentication/Interceptors.js (hunk 1 above)
Comment (keeganbrown) [reply modules/authentication/Interceptors.js]: for what its worth, `keySeq` also maintains the orders of the keys, whereas `keys` does not. Personally I've found `keySeq`, 'valueSeq`, and 'entrySeq` to behave much more practicably than their non-`seq` counterparts.

File: modules/authentication/services/index.js [hunk 3]
```diff
@@ -8,7 +8,7 @@ function checkAuth(store) {
     return store.getHydratedState()
       .then((state) => {
         const hasAllTokenInfoKeys = VALID_TOKEN_INFO_FIELDS.filter((field) => {
-          return state.authentication.tokenInfo[field];
+          return state.getIn(['authentication', 'tokenInfo', field]);
```
Comment (keeganbrown) [inline modules/authentication/services/index.js]: so we're using immutable for the root level state? I'm sure you already discussed the pros and cons of doing so, are we roughly aware of what those are?

Comment (leeallen337) [reply modules/authentication/services/index.js]: My understanding is that it's just more inline with the documentation around using Immutable.JS with Redux. It's [this](https://redux.js.org/recipes/using-immutable.js-with-redux#make-your-entire-redux-state-tree-an-immutablejs-object) page in particular but for some reason it's not loading the CSS for me so I'll just paste a screenshot

screen shot 2018-02-19 at 3 56 50 pm

I guess it's just that if you're already using Immutable.JS for all your reducers just might as well make your top level Immutable too

**Comment by leeallen337** [comment]
@keeganbrown Oh definitely on adding some basic `reselect` scaffolding. That will be coming in another pull request 👍

**Comment by leeallen337** [comment]
Thanks for the reviews @dkniffin @OzoneTank @keeganbrown 👍

**Pull request merged**